File: api/src/main/java/dev/akif/crud/CRUDDTOMapper.java
package dev.akif.crud;

import java.io.Serializable;

/**
 * Mapper to convert between DTOs, models and entities
 *
 * @param <I>  Id type of the data
 * @param <E>  Entity type of the data
 * @param <M>  Model type of the data
 * @param <D>  DTO type of the data
 * @param <CM> Create model type of the data
 * @param <UM> Update model type of the data
 * @param <CD> Create DTO type of the data
 * @param <UD> Update DTO type of the data
 */
public interface CRUDDTOMapper<
        I extends Serializable,
        E extends CRUDEntity<I, E>,
        M extends CRUDModel<I>,
        D extends CRUDDTO<I>,
        CM extends CRUDCreateModel<I, E>,
        UM extends CRUDUpdateModel<I, E>,
        CD extends CRUDCreateDTO<I, E, CM>,
        UD extends CRUDUpdateDTO<I, E, UM>> {
    /**
     * Mapper to convert from model to DTO
     *
     * @param model Model to convert
     * @return DTO built from given model
     */
    D modelToDTO(final M model);

    /**
     * Mapper to convert given create DTO into a create model
     *
     * @param createDTO Create DTO to convert
     * @return Create model built from this create DTO
     */
    CM createDTOToCreateModel(final CD createDTO);

    /**
     * Mapper to convert given update DTO into a update model
     *
     * @param updateDTO Update DTO to convert
     * @return Create model built from this update DTO
     */
    UM updateDTOToUpdateModel(final UD updateDTO);
}

File: api/src/main/kotlin/dev/akif/crud/CRUDDTOMapper.kt
package dev.akif.crud

import dev.akif.crud.common.Parameters
import java.io.Serializable

/**
 * Mapper to convert between DTOs, models and entities
 *
 * @param I  Id type of the data
 * @param M  Model type of the data which is a [CRUDModel]
 * @param D  DTO type of the data which is a [CRUDDTO]
 * @param CM Create model type of the data which is a [CRUDCreateModel]
 * @param UM Update model type of the data which is a [CRUDUpdateModel]
 * @param CD Create DTO type of the data which is a [CRUDCreateDTO]
 * @param UD Update DTO type of the data which is a [CRUDUpdateDTO]
 */
interface CRUDDTOMapper<
        I : Serializable,
        in M : CRUDModel<I>,
        out D : CRUDDTO<I>,
        out CM : CRUDCreateModel,
        out UM : CRUDUpdateModel,
        in CD : CRUDCreateDTO,
        in UD : CRUDUpdateDTO> {
    /**
     * Mapper to convert from model to DTO
     *
     * @param model Model to convert
     * @param parameters [Parameters] to use in conversion
     * @return DTO built from given model
     */
    fun modelToDTO(model: M, parameters: Parameters): D

    /**
     * Mapper to convert given create DTO to a create model
     *
     * @param createDTO Create DTO to convert
     * @param parameters [Parameters] to use in conversion
     * @return Create model built from this create DTO
     */
    fun createDTOToCreateModel(createDTO: CD, parameters: Parameters): CM

    /**
     * Mapper to convert given update DTO to a update model
     *
     * @param updateDTO Update DTO to convert
     * @param parameters [Parameters] to use in conversion
     * @return Create model built from this update DTO
     */
    fun updateDTOToUpdateModel(updateDTO: UD, parameters: Parameters): UM
}
